### PR TITLE
Enhance the FrameCaptureSystemComponent to support multiple active captures at the same time

### PR DESCRIPTION
**Pull request by rgba16f**
FrameCaptureBus now returns a capture id with each request. 
Calling code uses the capture id to identify when it's capture is complete.
Update any code that uses the affected API.

### CODE REVIEW

**Review by galibzon — APPROVED** [review]

**Review by gadams3 — CHANGES_REQUESTED** [review]

File: Gems/Atom/Tools/AtomToolsFramework/Code/Source/PreviewRenderer/PreviewRendererCaptureState.cpp [hunk 1]
```diff
@@ -27,16 +27,26 @@ namespace AtomToolsFramework
...
         {
-            AZ::Render::FrameCaptureNotificationBus::Handler::BusConnect();
-            AZ::TickBus::Handler::BusDisconnect();
+            m_frameCaptureId = m_renderer->StartCapture();
+            if (m_frameCaptureId != AZ::Render::FrameCaptureRequests::s_InvalidFrameCaptureId)
+            {
+                AZ::Render::FrameCaptureNotificationBus::Handler::BusConnect();
+                AZ::TickBus::Handler::BusDisconnect();
+            }
+            // if the start capture call fails the capture will be retried next tick.
```
[the remark this answers is not in the document]
Comment (gadams3) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Source/PreviewRenderer/PreviewRendererCaptureState.cpp:38]: Discussed offline. Changes for addressing the bus will be done in a follow-up PR

File: Gems/Atom/Tools/AtomToolsFramework/Code/Source/PreviewRenderer/PreviewRenderer.cpp [hunk 2]
```diff
@@ -66,6 +66,7 @@ namespace AtomToolsFramework
         pipelineDesc.m_renderSettings.m_multisampleState = AZ::RPI::RPISystemInterface::Get()->GetApplicationMultisampleState();
 
         m_renderPipeline = AZ::RPI::RenderPipeline::CreateRenderPipeline(pipelineDesc);
+        //m_renderPipeline->RemoveFromRenderTick();
         m_scene->AddRenderPipeline(m_renderPipeline);
         m_scene->Activate();
         AZ::RPI::RPISystemInterface::Get()->RegisterScene(m_scene);
```
[the remark this answers is not in the document]
Comment (rgba16f) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Source/PreviewRenderer/PreviewRenderer.cpp]: Removed that line.

**Review by gadams3 — APPROVED** [review]

File: Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuter.cpp [hunk 3]
```diff
@@ -106,6 +106,8 @@ namespace AZ
 
                 // Detect if we are able to continue merging.
                 {
+                    const AZStd::string_view scopeName = scope.GetId().GetStringView();
+                    const bool frameCaptureScope = scopeName.starts_with("FrameCapture_");
```
[the remark this answers is not in the document]
Comment (rgba16f) [reply Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuter.cpp:110]: I don't disagree that this approach is brittle. The scope solution would need implementations for every RHI though. Is there an existing way to add DX12 specific workarounds?

File: Gems/Atom/Feature/Common/Code/Source/FrameCaptureSystemComponent.cpp [hunk 4]
```diff
@@ -313,21 +316,45 @@ namespace AZ
...
+                capture->Reset();
+                return captureHandle;
+            }
+            else
+            {
+                AZStd::lock_guard<AZStd::shared_mutex> lock(m_handleLock); // take exclusive write lock as we may move CaptureState locations in memory
+                uint32_t captureIndex = aznumeric_cast<uint32_t>(m_allCaptures.size());
+                m_allCaptures.emplace_back(captureIndex);
+                return CaptureHandle(this, captureIndex);
             }
```
[the remark this answers is not in the document]
Comment (rgba16f) [reply Gems/Atom/Feature/Common/Code/Source/FrameCaptureSystemComponent.cpp]: Added suggested comments

File: Gems/Atom/Feature/Common/Code/Source/FrameCaptureSystemComponent.h [hunk 5]
```diff
@@ -35,39 +39,69 @@ namespace AZ
...
+            uint32_t CaptureScreenshotForWindow(const AZStd::string& filePath, AzFramework::NativeWindowHandle windowHandle) override;
+            uint32_t CaptureScreenshotWithPreview(const AZStd::string& outputFilePath) override;
+            uint32_t CapturePassAttachment(const AZStd::vector<AZStd::string>& passHierarchy, const AZStd::string& slotName, const AZStd::string& outputFilePath,
                 RPI::PassAttachmentReadbackOption option) override;
-            bool CapturePassAttachmentWithCallback(const AZStd::vector<AZStd::string>& passHierarchy, const AZStd::string& slotName
+            uint32_t CapturePassAttachmentWithCallback(const AZStd::vector<AZStd::string>& passHierarchy, const AZStd::string& slotName
                 , RPI::AttachmentReadback::CallbackFunction callback, RPI::PassAttachmentReadbackOption option) override;
 
         private:
+            struct CaptureState
```
[the remark this answers is not in the document]
Comment (rgba16f) [reply Gems/Atom/Feature/Common/Code/Source/FrameCaptureSystemComponent.h:52]: Added a small class header comment

File: Gems/Atom/Feature/Common/Code/Source/FrameCaptureSystemComponent.h [hunk 6]
```diff
@@ -35,39 +39,69 @@ namespace AZ
...
+                CaptureState(uint32_t captureId);
+                CaptureState(CaptureState&& rhs);
+                void Reset();
+                AZStd::shared_ptr<AZ::RPI::AttachmentReadback> m_readback;
+                AZStd::string m_outputFilePath;
+
+                AZStd::atomic<FrameCaptureResult> m_result = FrameCaptureResult::None; // use atomic as this is written and read on different threads.
+                AZStd::string m_latestCaptureInfo;
+            };
+            class CaptureHandle
```
[the remark this answers is not in the document]
Comment (rgba16f) [reply Gems/Atom/Feature/Common/Code/Source/FrameCaptureSystemComponent.h:66]: Added a small class header comment

File: Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuter.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rgba16f) [reply Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuter.cpp]: Removed this hack, it wasn't fixing the original issue (1 camera on DX12 always returns black data).

**Pull request merged**